### PR TITLE
feat(geneva-exporter): configurable event names and attribute-based table routing

### PR DESCRIPTION
**Pull request by ragumarimuthu-git**
## Description

Adds **attribute-based event/table-name routing** to the Geneva exporter. The exporter config now accepts optional `logs:` and `spans:` blocks, each with a `default_event_name` and an optional `event_name_mapping` that routes telemetry to different Geneva tables based on an attribute value.

### What's new

- **Separate, configurable event names** for logs and spans via `logs.default_event_name` / `spans.default_event_name` (default `Log` / `Span` when omitted). Replaces the previous unified `event_name`.
- **Attribute-based table routing** via `event_name_mapping`:
  - Logs can route on `event_name`, `resource_attribute`, `scope_attribute`, or `log_record_attribute`.
  - Spans can route on `resource_attribute`, `scope_attribute`, or `span_attribute`.
  - Reserved routing-key names `scope.name` / `scope.version` route on the instrumentation scope.
  - A `null` mapping value uses the source attribute value as the destination table name.
- Internal config types `LogConfig`/`SpanConfig` renamed to `LogsConfig`/`TracesConfig`; both sections are optional, aligning the exporter config with the geneva-uploader `GenevaClientConfig`.

### Config validation hardening (from review)

- `#[serde(deny_unknown_fields)]` on the nested config wrappers.
- Blank / whitespace-only `default_event_name` is rejected.
- Consistent deserialization error for `{ event_name: null }`.
- Added negative/edge-case tests (empty blocks, explicit null, unknown fields).

### Dependency

- Bumps the `geneva-uploader` dependency to `98aa6dc8717b372ab34ebe3247262379d4e33799` (full 40-char SHA) with `tls-rustls`, picking up the routing API. Both `Cargo.toml` and `Cargo.lock` are in sync.

### Example configs

- `otlp-geneva-set-table-name.yaml` and `otlp-geneva-set-routing-table-name.yaml` added/updated with placeholder endpoints/certs and accurate headers/comments.

## Testing

- `cargo build`/`test --workspace --features geneva-exporter --lib --bins --tests`: build OK, **4,781 passed, 0 failed, 8 ignored**.
- clippy clean, fmt applied.
- Release build (`--features geneva-exporter`) validated; Geneva exporter registered (`urn:microsoft:exporter:geneva`); example configs validate and route as configured.

### CODE REVIEW

**Comment by ragumarimuthu-git** [comment]
Addressed the Copilot review comments in `3799d81cb` (and earlier commits):

**`mod.rs` — routing-key deserializer `expecting()` message**
Updated the message to also list `event_name` in the map form, so it matches what `visit_map` actually accepts and deserialization errors are no longer misleading.

**`mod.rs` — missing test documentation**
Added `/// Scenario:` / `/// Guarantees:` doc comments (per `rust/otap-dataflow/AGENTS.md`) to the 16 newly-added tests that were missing them (the routing-key / mapping / conversion tests). Pre-existing tests untouched.

**`transformer.rs` — `test_utf8_multibyte_strings` doc format**
Reworded the doc comment into the required `Scenario:` / `Guarantees:` format.

**Already resolved in earlier commits:**
- YAML `event_name_mapping` indentation — fixed (2-space nesting) in `7dd3731e9`.
- Both changelog entries' empty `issues:` — now `issues: [3600]`.

Validation: `cargo fmt` clean; `cargo build`/`test --workspace --features geneva-exporter --lib --bins --tests` — build OK, tests pass, 0 failures.

File: rust/otap-dataflow/.chloggen/geneva-exporter-tls-rustls.yaml [hunk 1]
```diff
@@ -1,27 +0,0 @@
-# Use this changelog template to create an entry for release notes.
```
Comment (ThomsonTan) [inline rust/otap-dataflow/.chloggen/geneva-exporter-tls-rustls.yaml:1]: Why is this deleted?

Comment (ragumarimuthu-git) [reply rust/otap-dataflow/.chloggen/geneva-exporter-tls-rustls.yaml:1]: Good catch - this deletion was a mistake. This entry actually belongs to #3408 (the TLS rustls/SymCrypt switch), which is already merged to main, so it's an unrelated *pending* release note that should stay until the next CHANGELOG is generated. I've restored it from main in 2f9341cd4.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 2]
```diff
@@ -72,6 +73,351 @@ use otap_df_telemetry::common_attributes::{Outcome, SignalOutcomeAttributes};
...
+
+                while let Some(key) = map.next_key::<String>()? {
+                    if found_key.is_some() {
+                        return Err(de::Error::custom(
+                            "routing_key must have exactly one field, found multiple fields",
+                        ));
+                    }
+
+                    match key.as_str() {
+                        "event_name" => {
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:151]: Can we reject the map form for `event_name`? Today `{ event_name: some_attribute }` is accepted, but `some_attribute` is silently ignored and the exporter routes using the log record's built-in event name. This can make an incorrect configuration look valid while routing to the wrong table. Since the documented form is `routing_key: event_name`, I think we should support only that form and require `log_record_attribute` when the user wants to route by an attribute.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 3]
```diff
@@ -72,6 +73,351 @@ use otap_df_telemetry::common_attributes::{Outcome, SignalOutcomeAttributes};
...
+
+/// Deserializable wrapper for LogsEventNameMapping
+#[derive(Debug, Deserialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct LogsEventNameMappingConfig {
+    /// The routing key configuration (determines which attribute to route on)
+    pub routing_key: LogsEventNameRoutingKeyConfig,
+    /// Map of source values to destination table names.
+    /// A null value means the source value is used as the destination.
+    pub events: std::collections::HashMap<String, Option<String>>,
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:305]: The selected table name eventually becomes the uploader's `event=` query parameter, but `geneva-uploader/src/ingestion_service/uploader.rs` inserts `event_name` without URL-encoding it. This PR now allows configurable and attribute-derived table names, so values containing `&`, `#`, `%`, spaces, or non-ASCII characters can corrupt the upload URL. Can we fix `geneva-uploader` to encode `event_name` and update the dependency pin before exposing this configuration?

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs (hunk 2 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:151]: There are two ways to provide `routing_key`:

- **String form** - `routing_key: event_name`. When the user chooses `event_name`, no key/value pair is required because the routing value comes from the log record's built-in event name, so there is nothing extra to supply.
- **Map form** - `routing_key: { resource_attribute: <name> }` (or `scope_attribute` / `log_record_attribute`). In every case other than `event_name` a value **is** required, and it is validated as a non-empty attribute name.

So `event_name` is the only variant that carries no accompanying value; all attribute-based variants require one.

Comment (lalitb) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:151]: Probably, I was not clear above. My ask is only to fail configuration validation of this form:

```yaml
  routing_key: 
       event_name: MyTable
```

Currently, `MyTable` is silently ignored, and the exporter routes using the log record's built-in `event_name`. That is surprising because the configuration is accepted even though its value has no effect.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs (hunk 3 above)
Comment (lalitb) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:305]: Discussed offline with Raghu, this could be existing issue in geneva uploader, and so can be handled separately.

Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:305]: Yes, this is a valid check. This PR doesn't change the current uploader implementation logic - `event_name` is passed through to `geneva-uploader` exactly as before, and the missing URL-encoding lives in `geneva-uploader/src/ingestion_service/uploader.rs` (there's already a `// TODO - Maintain this as url-encoded...` there). Since the fix belongs upstream in `opentelemetry-rust-contrib`, we can look into it in a future `geneva-uploader` release and bump the pin then.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs (hunk 2 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:151]: Thanks for clarifying - done in ac2d9e53c. The map form now **fails** configuration validation:

```yaml
routing_key:
  event_name: MyTable   # rejected: "'event_name' does not take a value; use the string form `routing_key: event_name` instead of a map"
```

`event_name` is now only selectable via the string form `routing_key: event_name`. Attribute-based variants (`resource_attribute` / `scope_attribute` / `log_record_attribute`) still use the map form and require a non-empty value. Updated the `expecting()` message, the empty-map error, and the corresponding test (`test_logs_routing_key_event_name_map_form_is_rejected`). Full workspace tests + clippy pass.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 4]
```diff
@@ -72,6 +73,350 @@ use otap_df_telemetry::common_attributes::{Outcome, SignalOutcomeAttributes};
...
+
+        impl<'de> Visitor<'de> for RoutingKeyVisitor {
+            type Value = LogsEventNameRoutingKeyConfig;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                formatter.write_str(
+                    "either a string 'event_name' or a map with exactly one of: \
+                     'resource_attribute', 'scope_attribute', or 'log_record_attribute'",
+                )
+            }
```
[the remark this answers is not in the document]
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:116]: Addressed. The custom deserializer no longer accepts the map form `{ event_name: <any> }` at all - per later reviewer feedback it now **fails** validation (commit `ac2d9e53c`). The `expecting()` message was updated to accurately describe the accepted forms: the string `'event_name'` or a map with exactly one of `resource_attribute` / `scope_attribute` / `log_record_attribute`. So the message and the accepted forms are now consistent.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 5]
```diff
@@ -1232,6 +1630,1155 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_routing_key_event_name_variant() {
```
[the remark this answers is not in the document]
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:1732]: Fixed in `3799d81cb`. Added `/// Scenario:` / `/// Guarantees:` doc comments (per `rust/otap-dataflow/AGENTS.md:49-57`) to the newly-added tests that were missing them. The geneva_exporter module now has matching Scenario/Guarantees docs across its tests.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/azure_monitor_exporter/transformer.rs [hunk 6]
```diff
@@ -1352,7 +1352,7 @@ mod tests {
         }
     }
 
-    /// Test that UTF-8 multi-byte strings (emoji, CJK, etc.) survive the
+    /// Test UTF-8 multi-byte strings (emoji, CJK, etc.) survive the
     /// direct JSON serialization without corruption.
```
[the remark this answers is not in the document]
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/azure_monitor_exporter/transformer.rs]: Fixed in `3799d81cb`. `test_utf8_multibyte_strings` now uses the required `/// Scenario:` / `/// Guarantees:` format.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+              max_concurrent_uploads: 4
+              spans:
+                default_event_name: "CustomSpanTable"
+                # Example span routing configurations (similar to logs):
+                event_name_mapping:
+                 routing_key:
+                   resource_attribute: cluster
+                 events:
+                   MyTable: MySpanTable
+                   SpanTable1:
```
[the remark this answers is not in the document]
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml]: Fixed in `7dd3731e9`. The `event_name_mapping` block now uses consistent 2-space nesting matching the surrounding YAML.

File: rust/otap-dataflow/.chloggen/geneva-exporter-configurable-event-name.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# Changelog entry for Geneva exporter configurable event name feature
+
+change_type: enhancement
+
+component: pipeline
+
+note: "Geneva exporter now supports separate, configurable event names for logs and spans. Use `logs.default_event_name` and `spans.default_event_name` to control which Geneva tables telemetry is routed to. Defaults to 'Log' for logs and 'Span' for spans if not specified."
+
+issues: []
```
[the remark this answers is not in the document]
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/.chloggen/geneva-exporter-configurable-event-name.yaml]: Fixed in `5ad708b5b` - `issues:` is now `[3600]`.

File: rust/otap-dataflow/.chloggen/geneva-exporter-event-name-routing.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# Changelog entry for Geneva exporter attribute-based table routing
+
+change_type: enhancement
+
+component: pipeline
+
+note: "Geneva exporter now supports routing logs and spans to different Geneva tables based on an attribute value via `event_name_mapping`. Logs can route on `event_name`, `resource_attribute`, `scope_attribute`, or `log_record_attribute`; spans can route on `resource_attribute`, `scope_attribute`, or `span_attribute`. The reserved keys `scope.name` and `scope.version` can be used as the routing attribute name to route on the instrumentation scope."
+
+issues: []
```
[the remark this answers is not in the document]
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/.chloggen/geneva-exporter-event-name-routing.yaml]: Fixed in `5ad708b5b` - `issues:` is now `[3600]`.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:1177]: The new routing config has validation in two different places. For example, `events: {}` or a blank source key passes `--validate`, but later fails inside `GenevaClient::new()` when the pipeline starts.

Can we run the full routing validation here too? A config accepted by `--validate` should not fail later during pipeline startup.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 10]
```diff
@@ -72,6 +73,355 @@ use otap_df_telemetry::common_attributes::{Outcome, SignalOutcomeAttributes};
...
+
+/// Deserializable wrapper for LogsEventNameMapping
+#[derive(Debug, Deserialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct LogsEventNameMappingConfig {
+    /// The routing key configuration (determines which attribute to route on)
+    pub routing_key: LogsEventNameRoutingKeyConfig,
+    /// Map of source values to destination table names.
+    /// A null value means the source value is used as the destination.
+    pub events: std::collections::HashMap<String, Option<String>>,
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:305]: This PR exposes empty destination strings through the new `events` configuration, and the uploader treats them as passthrough just like null values:

```yaml
  Foo:
  Foo: ""
```

Only the null form is documented. Can we either reject empty and whitespace-only destinations, or document them as another supported passthrough form? I would prefer rejecting them because they are easy to enter by mistake.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 11]
```diff
@@ -313,12 +686,8 @@ impl GenevaExporter {
...
             role_instance: config.role_instance.clone(),
             msi_resource,
-            logs: LogsConfig {
-                default_event_name: None,
-            },
-            spans: TracesConfig {
-                default_event_name: None,
-            },
+            logs: logs_config,
+            spans: traces_config,
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs]: Most of the new tests stop after deserializing the local config types. Can we add one test for the complete conversion into `GenevaClientConfig`?

The test should check the logs and spans routing-key variants, event maps, and default table names. This is the adapter behavior owned by this PR, and it is where a logs/spans mix-up or incorrect `From` conversion could otherwise go unnoticed.

File: rust/otap-dataflow/.chloggen/geneva-exporter-configurable-event-name.yaml [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+issues: [3600]
+
+subtext: |
+  The configuration now accepts separate `logs` and `spans` blocks with independent `default_event_name` settings:
+    logs:
+      default_event_name: "CustomLogTable"
+    spans:
+      default_event_name: "CustomSpanTable"
+  Each field is optional and will default to "Log" or "Span" respectively if omitted. This replaces the previous unified `event_name` configuration, providing clearer separation of concerns between log and trace signal types.
```
Comment (lalitb) [inline rust/otap-dataflow/.chloggen/geneva-exporter-configurable-event-name.yaml]: > This replaces the previous unified `event_name` configuration

There is no previous unified `event_name` setting in the Geneva exporter config on `main`, so this sounds like a migration users need to make when none exists.

Can we remove this sentence and describe `logs.default_event_name` and `spans.default_event_name` as new optional settings?

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-table-name.yaml [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+                password: "your-cert-password"
+              max_buffer_size: 1000
+              max_concurrent_uploads: 4
+              logs:
+                default_event_name: "CustomLogTable"
+              spans:
+                default_event_name: "CustomSpanTable"
+
+        connections:
+          # Shared receiver -> type router
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-table-name.yaml]:
```suggestion
          # Shared receiver -> Geneva exporter
```

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+                # Or use log_record_attribute:
+                # event_name_mapping:
+                #  routing_key:
+                #    log_record_attribute: custom_event_name  # Note: event_name at this level will not work, it has to be log_record_attribute
+                #  events:
+                #    MyTable: MyLogTable
+                #    LogTable1: MyLogTable
+
+              spans:
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml]: This `spans:` block has no active fields, so it parses as null and has no effect. It is easy to mistake it for an active span configuration.

Can we remove the empty block and keep this explanation as normal comments? The working span configuration is already shown in the trace exporter below.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+              - logs
+              - traces
+            config: {}
+
+          geneva-exporter-logs:
+            type: "urn:microsoft:exporter:geneva"
+            config:
+              endpoint: "endpoint"
+              environment: "environment"
+              account: "account"
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml]: nit -   These placeholders look like ordinary values, but copying the example gives an invalid Geneva endpoint and fails later with a confusing initialization error.

Can we use clearly marked placeholders such as `https://your-geneva-endpoint.example`, `your-environment`, and `your-account`, matching the main Geneva example?

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+                #     MyTable: MyLogTable
+                #     LogTable1: MyLogTable
+
+          geneva-exporter-traces:
+            type: "urn:microsoft:exporter:geneva"
+            config:
+              endpoint: "endpoint"
+              environment: "environment"
+              account: "account"
+              namespace: "namespace"
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml]: nit -   These placeholders look like ordinary values, but copying the example gives an invalid Geneva endpoint and fails later with a confusing initialization error.

  Can we use clearly marked placeholders such as `https://your-geneva-endpoint.example`, `your-environment`, and `your-account`, matching the main Geneva example?

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 17]
```diff
@@ -72,6 +73,355 @@ use otap_df_telemetry::common_attributes::{Outcome, SignalOutcomeAttributes};
 /// The URN for the Geneva exporter
 pub const GENEVA_EXPORTER_URN: &str = "urn:microsoft:exporter:geneva";
 
+/// Deserializable wrapper for LogsEventNameRoutingKey
+#[derive(Debug, Clone)]
+pub enum LogsEventNameRoutingKeyConfig {
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:78]: Non-blocking follow-up: these local types duplicate the routing model from `geneva-uploader` only to add serde support. This leaves the YAML format in this repository while the validation and behavior live in the uploader.

Can we track ( using an issue) adding optional serde support to the uploader routing types, then deserialize those types directly here? That would keep the wire format and its validation together and prevent the two definitions from drifting.

Same for `LogsEventNameMappingConfig`, `SpansEventNameRoutingKeyConfig`, `SpansEventNameMappingConfig`.

**Comment by lalitb** [comment]
The overall direction looks good for this PR.  I left a few straightforward comments on this PR.  I also noticed two performance follow-ups in the `opentelemetry-rust-contrib` Geneva uploader while reviewing this PR, that would be good to capture in a tracking issue:

- [routing.rs:363](https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/d2eeff7e91f39889405320cdc91c1c18332fcac9/opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/routing.rs#L363) -  `EventName` log routing copies the borrowed event name only to perform the map lookup.
- [routing.rs:164](https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/d2eeff7e91f39889405320cdc91c1c18332fcac9/opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/routing.rs#L164) - Span passthrough routing copies `source_value`, even though it already borrows from the span.

We should be able to preserve the borrow in these two paths? The log-attribute path is different and may still need ownership because the pdata view value is a temporary.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:1177]: Fixed in 84eb9e219. Events-map validation now runs during deserialization (via a serde `try_from` wrapper), so both the `--validate` path and pipeline startup reject an empty `events` map and blank/whitespace source keys up-front, instead of only failing later inside `GenevaClient::new()`.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs (hunk 10 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:305]: Fixed in 84eb9e219. Empty/whitespace destination values are now rejected during deserialization. A `null` destination remains valid (passthrough = keep the source event name), but `Foo: \\` is no longer silently treated as passthrough.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs (hunk 11 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs]: Done in 84eb9e219. Extracted `Config::to_geneva_client_config` and added `test_config_to_geneva_client_config_full_adapter`, which asserts every `GenevaClientConfig` field (endpoint, environment, account, namespace, region, version, tenant, role_name, role_instance, auth_method, msi_resource, logs, spans, obo_event_map).

File: rust/otap-dataflow/.chloggen/geneva-exporter-configurable-event-name.yaml (hunk 12 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/.chloggen/geneva-exporter-configurable-event-name.yaml]: Fixed in 84eb9e219. Removed the `replaces the previous unified event_name` wording and reworded the entry to describe these as new optional settings.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-table-name.yaml (hunk 13 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-table-name.yaml]: Fixed in 84eb9e219. Updated the connection comment to `# Shared receiver -> Geneva exporter`.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml (hunk 14 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml]: Fixed in 84eb9e219. Removed the empty `spans:` block (which parsed as null) and replaced it with commented-out example span mapping.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml (hunk 15 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml]: Fixed in 84eb9e219. Replaced the bare placeholders with clearly-marked example values (e.g. `https://your-geneva-endpoint.example`, `your-environment`, `your-account`, ...) in both exporter blocks.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml (hunk 16 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/otlp-geneva-set-routing-table-name.yaml]: Fixed in 84eb9e219 as part of the same placeholder cleanup for the spans block.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs (hunk 17 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:78]: Noted as a follow-up. This is non-blocking; the local routing types intentionally mirror the uploader model for now, and adding serde derives to the uploader types to remove the duplication can be handled separately in a future geneva_uploader change.

**Comment by ragumarimuthu-git** [comment]
Thanks! The inline comments have been addressed in `84eb9e219` (validation moved into deserialization, `to_geneva_client_config` adapter + full-adapter test, empty-destination rejection, and the doc/YAML nits).

For the two performance follow-ups in the `opentelemetry-rust-contrib` Geneva uploader, I've captured them in a tracking issue: **open-telemetry/opentelemetry-rust-contrib#718**. It covers both `routing.rs:363` (EventName log routing copying the borrowed event name for the map lookup) and `routing.rs:164` (span passthrough copying `source_value`), and notes that the log-attribute path may still need ownership because the pdata view value is a temporary. These are non-blocking for this PR.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs (hunk 17 above)
Comment (ragumarimuthu-git) [reply rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:78]: Tracked as a follow-up: **open-telemetry/opentelemetry-rust-contrib#719** — add optional serde support to the uploader routing types so this exporter can deserialize them directly instead of maintaining local duplicates, keeping the wire format and its validation together. Non-blocking for this PR.

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 18]
```diff
@@ -72,6 +73,430 @@ use otap_df_telemetry::common_attributes::{Outcome, SignalOutcomeAttributes};
...
+            {
+                match value {
+                    "event_name" => Ok(LogsEventNameRoutingKeyConfig::EventName),
+                    _ => Err(de::Error::unknown_variant(
+                        value,
+                        &[
+                            "event_name",
+                            "resource_attribute",
+                            "scope_attribute",
+                            "log_record_attribute",
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:130]: nit - this error message is confusing, as the only valid string form here is "event_name".

File: rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs [hunk 19]
```diff
@@ -72,6 +73,430 @@ use otap_df_telemetry::common_attributes::{Outcome, SignalOutcomeAttributes};
...
+                                    "log_record_attribute",
+                                ],
+                            ));
+                        }
+                    }
+                }
+
+                let non_empty = |field: &str, value: Option<String>| -> Result<String, M::Error> {
+                    let value = value.unwrap_or_default();
+                    if value.trim().is_empty() {
```
Comment (lalitb) [inline rust/otap-dataflow/crates/contrib-nodes/src/exporters/geneva_exporter/mod.rs:191]: nit: this checks the trimmed value, but stores the original value with spaces.

For example, `" service.name "` passes validation but will not match the usual `service.name` attribute. Can we reject leading or trailing whitespace instead of accepting it?

**Review by lalitb — APPROVED** [review]
LGTM, with nit comments. Thanks for patience in resolving all comments.

@manishgoel3 FYI, this PR and #3579 now change the same Geneva exporter config and `to_geneva_client_config()` adapter. After this PR merges, #3579 will need to rebase and preserve both agent-fed authentication and the new logs/spans routing configuration. And also need to update the `ConfigBuilder` and add a test covering both features together.

**Comment by utpilla** [comment]
Towards #3375
